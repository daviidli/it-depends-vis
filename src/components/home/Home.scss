.iconContainer {
	background-color: #e74c3c;
	border-radius: 100em;
	width: 150;
	height: 150px;
	padding: 18px 10px;
	margin: 6em;
	margin-bottom: 2em;
}

.repoInput {
	margin: 6em;
	margin-top: 4em;
	max-width: 40em;
	border-radius: 100em;
	border: none;

	.ant-input-suffix {
		visibility: hidden;
	}
}

.repoIcon {
	margin-top: 6px;
}

.about {
	max-width: 40em;
}
