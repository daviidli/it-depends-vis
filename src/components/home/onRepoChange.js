import axios from 'axios';
import { pipe, prop, identity } from 'ramda';
import {
	fork, attemptP, mapRej, map
} from 'fluture/index';
import { push } from 'connected-react-router';
import store from '../../store/store';
import { setData, setError } from '../../actions/actions';
import { dispatch, dispatchThunk } from '../../utils/reduxUtils';
import routes from '../../constants/routes.json';

const goHome = dispatchThunk(store, push(routes.HOME));
const goToVis = dispatchThunk(store, push(routes.VIS));

const dispatchData = pipe(prop('data'), setData, dispatch(store));
const dispatchError = pipe(setError, dispatch(store));

const putRequest = repo => attemptP(() => axios.put(`http://localhost:8080/init?url=${repo}`));

const onRepoChange = pipe(
	putRequest,
	map(pipe(dispatchData, goToVis)),
	mapRej(pipe(dispatchError, goHome)),
	fork(identity)(identity)
);

export default onRepoChange;
