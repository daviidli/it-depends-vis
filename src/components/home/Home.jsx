import React, { useEffect } from 'react';
import PropTypes from 'prop-types';
import { Row, Typography, Input } from 'antd';
import { GoRepo } from 'react-icons/go';
import { GiThink } from 'react-icons/gi';
import { ifElse, equals, identity } from 'ramda';
import { useHistory } from 'react-router';
import description from '../../constants/description.json';
import onRepoChange from './onRepoChange';
import routes from '../../constants/routes.json';

import './Home.scss';

const Home = ({ repo, setRepo }) => {
	const history = useHistory();

	useEffect(() => {
		ifElse(
			equals(''),
			identity,
			x => {
				history.push(routes.LOADING);
				onRepoChange(x);
			}
		)(repo);
	}, [repo]);

	return (
		<>
			<Row justify="center">
				<div className="iconContainer">
					<GiThink size={128} color="#fff" />
				</div>
			</Row>
			<Row justify="center">
				<Typography.Title>it depends</Typography.Title>
			</Row>
			<Row justify="center">
				<Input.Search
					className="repoInput"
					size="large"
					placeholder="Repository"
					prefix={
						<GoRepo size={24} color="grey" className="repoIcon" />
					}
					onSearch={setRepo}
				/>
			</Row>
			<Row justify="center">
				<Typography.Title level={2}>About</Typography.Title>
			</Row>
			<Row justify="center">
				<div className="about">
					<Typography.Text>{ description.about }</Typography.Text>
				</div>
			</Row>
		</>
	);
};

Home.propTypes = {
	repo: PropTypes.string.isRequired,
	setRepo: PropTypes.func.isRequired
};

export default Home;
