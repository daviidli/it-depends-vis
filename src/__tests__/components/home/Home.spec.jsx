import React from 'react';
import { shallow, mount } from 'enzyme';
import toJson from 'enzyme-to-json';
import { useHistory } from 'react-router';
import { Input } from 'antd';
import Home from '../../../components/home/Home';
import onRepoChange from '../../../components/home/onRepoChange';

// jest error: https://jestjs.io/docs/en/manual-mocks#mocking-methods-which-are-not-implemented-in-jsdom
Object.defineProperty(window, 'matchMedia', {
	writable: true,
	value: jest.fn().mockImplementation(query => ({
		matches: false,
		media: query,
		onchange: null,
		addListener: jest.fn(), // deprecated
		removeListener: jest.fn(), // deprecated
		addEventListener: jest.fn(),
		removeEventListener: jest.fn(),
		dispatchEvent: jest.fn()
	}))
});

jest.mock('react-router');
jest.mock('../../../components/home/onRepoChange');

const defaultProps = {
	repo: '',
	setRepo: jest.fn()
};

const setup = (props = defaultProps) => {
	const wrapper = shallow(<Home {...props} />);
	return {
		wrapper,
		input: wrapper.find(Input.Search)
	};
};

const setupMount = (props = defaultProps) => {
	const wrapper = mount(<Home {...props} />);
	return { wrapper };
};

describe('Home component', () => {
	beforeEach(() => {
		jest.clearAllMocks();
	});

	it('should match snapshot', () => {
		useHistory.mockImplementation(jest.fn());
		const { wrapper } = setup();
		expect(toJson(wrapper)).toMatchSnapshot();
	});

	it('should not trigger effect as repo is empty string', done => {
		const mockPush = jest.fn();
		useHistory.mockImplementation(() => ({ push: mockPush }));
		onRepoChange.mockImplementation(jest.fn());
		setupMount();
		expect(mockPush).toHaveBeenCalledTimes(0);
		setImmediate(() => {
			expect(mockPush).toHaveBeenCalledTimes(0);
			expect(onRepoChange).toHaveBeenCalledTimes(0);
			done();
		});
	});

	it('should trigger effect', done => {
		const mockPush = jest.fn();
		useHistory.mockImplementation(() => ({ push: mockPush }));
		onRepoChange.mockImplementation(jest.fn());
		setupMount({ repo: 'repo1', setRepo: jest.fn() });
		setImmediate(() => {
			expect(mockPush).toHaveBeenCalledTimes(1);
			expect(mockPush).toHaveBeenCalledWith('/loading');
			expect(onRepoChange).toHaveBeenCalledTimes(1);
			done();
		});
	});

	it('should call setRepo on search', () => {
		useHistory.mockImplementation(jest.fn());
		const mockSetRepo = jest.fn();
		const { input } = setup({ repo: '', setRepo: mockSetRepo });
		input.simulate('search', 'repo1');
		expect(mockSetRepo).toHaveBeenCalledTimes(1);
		expect(mockSetRepo).toHaveBeenCalledWith('repo1');
	});
});
