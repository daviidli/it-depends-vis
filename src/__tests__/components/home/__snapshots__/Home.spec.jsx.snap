// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Home component should match snapshot 1`] = `
<Fragment>
  <Row
    gutter={0}
    justify="center"
  >
    <div
      className="iconContainer"
    >
      <GiThink
        color="#fff"
        size={128}
      />
    </div>
  </Row>
  <Row
    gutter={0}
    justify="center"
  >
    <Title>
      it depends
    </Title>
  </Row>
  <Row
    gutter={0}
    justify="center"
  >
    <Search
      className="repoInput"
      enterButton={false}
      onSearch={[MockFunction]}
      placeholder="Repository"
      prefix={
        <GoRepo
          className="repoIcon"
          color="grey"
          size={24}
        />
      }
      size="large"
    />
  </Row>
  <Row
    gutter={0}
    justify="center"
  >
    <Title
      level={2}
    >
      About
    </Title>
  </Row>
  <Row
    gutter={0}
    justify="center"
  >
    <div
      className="about"
    >
      <Text>
        Developers often encounter trouble traversing the information space of a large software system with extensive revision history. Looking through a list of commits is time-consuming, and fails to convey information in a form which efficiently communicates information about the system. it depends is a visualization which makes it easier for developers and managers to understand the dependencies between entities in a software system across commits.
      </Text>
    </div>
  </Row>
</Fragment>
`;
